pkg/virt-controller/watch/vm.go

```diff
@@ -576,6 +577,73 @@ func (c *VMController) updatePVCMemoryDumpAnnotation(vm *virtv1.VirtualMachine)
 	return nil
 }
 
+func (c *VMController) saveCurrentVMICPUCsPatch(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
+
+	vmiCopy := vmi.DeepCopy()
+	vmiCopy.Status.CurrentCPUTopology = vmi.Spec.Domain.CPU
+
+	oldJson, err := json.Marshal(vmiCopy.Status.CurrentCPUTopology)
+	if err != nil {
+		return err
+	}
+
+	patchOps := fmt.Sprintf(`{ "op": "add", "path": "/status/currentCPUTopology", "value": %s}`, string(oldJson))
+	patchBytes := controller.GeneratePatchBytes([]string{patchOps})
+
+	if len(patchBytes) == 0 {
+		return nil
+	}
+
+	_, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, patchBytes, &v1.PatchOptions{})
+	if err != nil {
+		return err
+	}
+
+	var cpu virtv1.CPU
+	cpu.Sockets = vm.Spec.Template.Spec.Domain.CPU.Sockets
+	cpu.Cores = vm.Spec.Template.Spec.Domain.CPU.Cores
+	cpu.Threads = vm.Spec.Template.Spec.Domain.CPU.Threads
+	cpu.MaxSockets = vmi.Spec.Domain.CPU.MaxSockets
```

Comment:
> The VM controller will only update the vmi.spec.domain.cpu.sockets and the VMI controller will validate if this is allowed (by checking that maxSockets is already present in the vmi spec), and will save the currentCPUTopology to start the hotplug process.

How would the vmi controller be able to save `currentCPUTopology` if `vmi.spec.domain.cpu.sockets` is already mutated?

Reply:
@vladikr Hi, regarding
```
* hot plug will still be possible if the VMI is running without a VM
```
Today updating the VMI sockets is possible only with one of kubevirt system accounts. Would you like me to allow this field to be changed by the user?

Reply:
@iholder101 on update we normally have both the old and the new objects... what am I missing?

Reply:
Sorry Vladik. I was too tired writing the previous comment :) You're right.

I think what you suggest here makes a lot of sense. It's important that this would be available for VMIs too if possible.

One thing that I still don't fully understand: when you say `can only start if the MaxSockets is already present on the VMI`, how can it not be present? If a VM is being created with LiveUpdateFeatures, the VMI would be created with maxSockets, right?